qiime2/core/cache.py

```diff
@@ -0,0 +1,511 @@
+        """
+        referenced_pools = set()
+        referenced_data = set()
+
+        # Walk over keys and track all pools and data referenced
+        # This needs to be locked so we ensure that we don't have other threads
+        # or processes writing refs that we don't see leading to us deleting
+        # their data
+        with self.lock:
+            for key in os.listdir(self.keys):
+                loaded_key = yaml.safe_load(open(self.keys / key))
+                referenced_pools.add(loaded_key['pool'])
+                referenced_data.add(loaded_key['data'])
+
+            # Since each key has at most a pool or data, we will end up with a
+            # None in at least one of these sets. We don't want it
+            referenced_pools.discard(None)
+            referenced_data.discard(None)
+
+            # Walk over pools and remove any that were not refered to by keys
+            # while tracking all data within those that were referenced
+            for pool in os.listdir(self.pools):
+                if pool not in referenced_pools:
+                    shutil.rmtree(self.pools / pool)
+                else:
+                    for data in os.listdir(self.pools / pool):
+                        referenced_data.add(data)
+
+            # Add references to data in process pools
+            for process_pool in os.listdir(self.process):
+                # Pick the creation time out of the pool name of format
+                # {pid}-time@user
+                create_time = float(process_pool.split('-')[1].split('@')[0])
+
+                if time.time() - create_time >= self.process_pool_lifespan:
+                    shutil.rmtree(self.process / process_pool)
+                else:
+                    for data in os.listdir(self.process / process_pool):
+                        referenced_data.add(data)
+
+            # Walk over all data and remove any that was not referenced
+            for data in os.listdir(self.data):
+                # If this assert is ever tripped something real bad happened
+                assert is_uuid4(data)
+
+                if data not in referenced_data:
+                    shutil.rmtree(self.data / data, ignore_errors=True)
+
+    def save(self, ref, key):
+        """Create our key then create our data. Returns a version of the data
+        backed by the key in the cache
+        """
+        # Create the key before the data, this is so that if another thread or
+        # process is running garbage collection it doesn't see our unkeyed data
+        # and remove it leaving us with a dangling reference and no data
+        with self.lock:
+            self._register_key(key, str(ref.uuid))
+
+        _copy_to_data(self, ref, self.data)
+
+        # We want this ref to also be backed by the proces pool so for the
+        # duration of this process we still have access to it even if we delete
+        # the key
+        self.process_pool.save(ref)
```

Comment:
@Oddant1, we probably want this to happen in `load()`, since you could end up with the same issue if you loaded a key, and then deleted it. (Also, since save falls through load, you get two birds with one stone).

Reply:
Good idea